<a href="https://colab.research.google.com/github/srdataml-droid/AI-diploma/blob/main/week2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("kalusamuel/lagos-weather-dataset")

print("Path to dataset files:", path)

100%|██████████| 386k/386k [00:00<00:00, 725kB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/kalusamuel/lagos-weather-dataset/versions/2


In [ ]:
path

'/root/.cache/kagglehub/datasets/kalusamuel/lagos-weather-dataset/versions/2'

In [ ]:
ldf =  pd.read_csv(r'/root/.cache/kagglehub/datasets/kalusamuel/lagos-weather-dataset/versions/2/lagos_all_years.csv')

In [ ]:
ldf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8676 entries, 0 to 8675
Data columns (total 34 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   datetime          7125 non-null   object 
 1   tempmax           8085 non-null   float64
 2   tempmin           8085 non-null   float64
 3   temp              7853 non-null   float64
 4   feelslikemax      8085 non-null   float64
 5   feelslikemin      8085 non-null   float64
 6   feelslike         7853 non-null   float64
 7   dew               7853 non-null   float64
 8   humidity          7853 non-null   float64
 9   precip            7853 non-null   float64
 10  precipprob        7853 non-null   float64
 11  precipcover       8676 non-null   float64
 12  preciptype        3736 non-null   object 
 13  snow              3928 non-null   float64
 14  snowdepth         3923 non-null   float64
 15  windgust          5876 non-null   float64
 16  windspeed         8078 non-null   float64


In [ ]:
ldf.describe()

,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,precip,precipprob,...,sealevelpressure,cloudcover,visibility,solarradiation,solarenergy,uvindex,severerisk,moonphase,windspeedmax,windspeedmin
count,8085.000000,8085.000000,7853.000000,8085.000000,8085.000000,7853.000000,7853.000000,7853.000000,7853.000000,7853.000000,...,7664.000000,7853.000000,7651.000000,5388.000000,5388.000000,5388.000000,1362.000000,8310.000000,366.00000,3.660000e+02
mean,30.324366,23.770860,27.433949,36.054657,24.260940,30.371323,23.977550,82.685916,5.034344,43.282822,...,1012.014770,61.281230,8.844504,195.742743,16.902116,7.031180,46.098385,0.483118,20.39918,5.867443e+06
std,5.807219,4.508929,1.601799,7.695888,5.151739,3.522264,1.934448,7.354521,18.059178,49.549896,...,1.813524,17.515913,2.489910,57.873993,5.002004,1.964733,26.326035,0.288559,11.51725,3.505645e+07
min,0.000000,0.000000,21.400000,0.000000,0.000000,21.400000,8.200000,33.800000,0.000000,0.000000,...,1005.900000,0.000000,0.100000,0.000000,0.000000,0.000000,10.000000,0.000000,0.00000,0.000000e+00
25%,29.000000,23.100000,26.200000,33.600000,23.100000,27.500000,23.500000,80.200000,0.000000,0.000000,...,1010.700000,50.900000,8.000000,153.100000,13.200000,6.000000,30.000000,0.250000,13.00000,0.000000e+00
50%,31.100000,24.100000,27.400000,37.400000,24.100000,30.200000,24.200000,83.700000,0.000000,0.000000,...,1011.900000,59.100000,9.200000,204.150000,17.600000,7.000000,60.000000,0.500000,16.60000,0.000000e+00
75%,33.100000,25.600000,28.600000,40.600000,25.600000,32.900000,25.000000,86.800000,2.000000,100.000000,...,1013.300000,71.400000,10.000000,240.300000,20.800000,9.000000,60.000000,0.750000,24.10000,3.600000e+00
max,56.000000,34.600000,34.600000,56.000000,45.100000,45.100000,66.500000,100.000000,299.000000,100.000000,...,1017.300000,100.000000,31.300000,303.200000,26.300000,10.000000,100.000000,0.980000,82.30000,2.147484e+08


In [ ]:
ldf.tail
 (20)

,datetime,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,precip,...,severerisk,sunrise,sunset,moonphase,conditions,description,icon,stations,windspeedmax,windspeedmin
8656,2024-06-14,27.0,24.0,26.4,30.7,24.0,28.5,24.4,89.0,21.0,...,60.0,2024-06-14T06:32:05,2024-06-14T19:01:44,0.25,"Rain, Partially cloudy",Partly cloudy throughout the day with afternoo...,rain,"DNMM,65201099999",NaN,NaN
8657,2024-06-15,30.7,24.5,26.8,38.1,24.5,29.0,24.4,87.1,2.0,...,10.0,2024-06-15T06:32:16,2024-06-15T19:01:58,0.29,"Rain, Partially cloudy",Partly cloudy throughout the day with afternoo...,rain,"DNMM,remote,65201099999,VRJT5",NaN,NaN
8658,2024-06-16,30.0,24.6,26.9,36.3,24.6,28.7,23.4,81.3,3.3,...,30.0,2024-06-16T06:32:28,2024-06-16T19:02:12,0.32,"Rain, Overcast",Cloudy skies throughout the day with a chance ...,rain,"remote,65201099999,VRJT5",NaN,NaN
8659,2024-06-17,32.0,24.9,27.8,41.3,24.9,31.5,24.6,82.8,14.0,...,30.0,2024-06-17T06:32:40,2024-06-17T19:02:26,0.35,"Rain, Partially cloudy",Partly cloudy throughout the day with rain.,rain,"DNMM,remote,VRJT5,65201099999",NaN,NaN
8660,2024-06-18,31.0,25.0,27.6,45.9,25.0,31.3,25.0,86.5,2.4,...,30.0,2024-06-18T06:32:52,2024-06-18T19:02:40,0.38,"Rain, Partially cloudy",Partly cloudy throughout the day with rain cle...,rain,"DNMM,remote,65201099999,VRJT5",NaN,NaN
8661,2024-06-19,32.0,25.2,28.6,40.2,25.2,33.1,25.2,82.4,0.0,...,30.0,2024-06-19T06:33:05,2024-06-19T19:02:53,0.42,Partially cloudy,Partly cloudy throughout the day.,partly-cloudy-day,"DNMM,65201099999,VRJT5",NaN,NaN
8662,2024-06-20,30.0,25.1,28.0,36.2,25.1,32.2,24.9,83.5,0.0,...,30.0,2024-06-20T06:33:18,2024-06-20T19:03:07,0.45,Partially cloudy,Partly cloudy throughout the day.,partly-cloudy-day,"DNMM,remote,65201099999,VRJT5",NaN,NaN
8663,2024-06-21,31.0,24.0,26.7,39.1,24.0,29.2,24.8,89.5,41.0,...,30.0,2024-06-21T06:33:31,2024-06-21T19:03:20,0.48,"Rain, Partially cloudy",Partly cloudy throughout the day with late aft...,rain,"DNMM,65201099999",NaN,NaN
8664,2024-06-22,31.0,24.0,27.1,39.1,24.0,30.5,24.9,88.5,6.0,...,30.0,2024-06-22T06:33:44,2024-06-22T19:03:32,0.50,"Rain, Partially cloudy",Partly cloudy throughout the day with early mo...,rain,"DNMM,65201099999",NaN,NaN
8665,2024-06-23,31.0,25.0,28.1,40.6,25.0,32.4,25.3,85.4,1.0,...,30.0,2024-06-23T06:33:57,2024-06-23T19:03:45,0.55,"Rain, Partially cloudy",Partly cloudy throughout the day with afternoo...,rain,"DNMM,remote,65201099999",NaN,NaN


In [ ]:
ldf.windspeedmax.unique()

array([ nan, 13. , 14.8, 16.6, 18.4, 12.7,  9.4, 11.2, 20.4, 33.5, 22.3,
       29.5, 13.4, 40.7, 42.5, 14.7, 16.7, 14.6, 25.9, 18.3, 44.3, 55.4,
       20.5, 38. , 27.7, 31.3, 48.2, 13.9, 52. , 16.5, 64.8, 51.8, 16.4,
       63. , 46.4,  9.5, 24.1, 70.2,  0. ,  7.6, 12.9, 11.6, 18. , 27.5,
       22.2, 13.1, 28.6, 35.3, 27.1, 32.8, 28.3,  5.4, 37.1, 77.8, 21.9,
       46.5, 30.9, 54.4, 23.6, 18.2,  7.7, 21.8, 82.3])

In [ ]:
ldf.windspeedmin.unique()

array([           nan, 1.90000000e+00, 0.00000000e+00, 1.00000000e-01,
       2.10000000e+00, 3.60000000e+00, 6.00000000e-01, 2.00000000e-01,
       5.30000000e+00, 5.40000000e+00, 3.50000000e+00, 1.80000000e+00,
       1.00000000e+00, 5.20000000e+00, 1.12000000e+01, 3.00000000e-01,
       2.20000000e+00, 1.48000000e+01, 2.14748365e+08, 7.00000000e-01,
       7.40000000e+00, 3.80000000e+00, 7.60000000e+00, 4.00000000e-01,
       2.00000000e+00, 3.40000000e+00])

In [ ]:
ldf.isna().sum()

,0
datetime,1551
tempmax,591
tempmin,591
temp,823
feelslikemax,591
feelslikemin,591
feelslike,823
dew,823
humidity,823
precip,823


In [ ]:
# 1. Drop rows with missing datetime (Crucial: 1,551 rows removed)
ldf = ldf.dropna(subset=['datetime']).reset_index(drop=True)

In [ ]:
# 2. Convert datetime formats safely
ldf['datetime'] = pd.to_datetime(ldf['datetime'])
ldf['sunrise'] = pd.to_datetime(ldf['sunrise'], errors='coerce')
ldf['sunset'] = pd.to_datetime(ldf['sunset'], errors='coerce')

In [ ]:
ldf = ldf.drop(columns=['windspeedmax', 'windspeedmin'], errors='ignore')

In [ ]:
# 4. Fill rain/snow categorical and numeric gaps
ldf['preciptype'] = ldf['preciptype'].fillna('none')
ldf['snow'] = ldf['snow'].fillna(0.0)
ldf['snowdepth'] = ldf['snowdepth'].fillna(0.0)

In [ ]:
# 5. Fill critical core metrics using time-based interpolation
# (Works beautifully now that missing datetimes are gone)
interpolate_cols = [
    'tempmax', 'tempmin', 'temp', 'feelslikemax', 'feelslikemin', 'feelslike',
    'dew', 'humidity', 'precip', 'precipprob', 'windspeed', 'winddir',
    'sealevelpressure', 'cloudcover', 'visibility', 'windgust'
]
ldf[interpolate_cols] = ldf[interpolate_cols].interpolate(method='linear').bfill().ffill()


In [ ]:
# 6. Handle Solar, Risk, and Astronomic constants
ldf['severerisk'] = ldf['severerisk'].fillna(0.0)
ldf[['solarradiation', 'solarenergy', 'uvindex']] = ldf[['solarradiation', 'solarenergy', 'uvindex']].fillna(0.0)
ldf['moonphase'] = ldf['moonphase'].interpolate(method='linear').bfill().ffill()

In [ ]:
#  Fill text descriptions and metadata
meta_cols = ['conditions', 'description', 'icon', 'stations']
ldf[meta_cols] = ldf[meta_cols].fillna('unknown')

In [ ]:
print("Remaining Nulls:", l
      df.isna().sum().sum())

Remaining Nulls: 732


In [ ]:
# 1. Fill the missing text values with a temporary label
ldf['preciptype'] = ldf['preciptype'].fillna('clear')

In [ ]:

# 2. Explode the column into binary 1s and 0s
precipitation_dummies = pd.get_dummies(ldf['preciptype'], prefix='is')

In [ ]:
ldf = pd.concat([ldf, precipitation_dummies], axis=1)
ldf = ldf.drop(columns=['preciptype'])

In [ ]:
# Drop columns that are text-heavy or purely administrative
ldf = ldf.drop(columns=['description', 'icon', 'stations'])


In [ ]:
# Convert text strings directly into clean numeric ID tracking codes (0, 1, 2, 3...)
ldf['conditions'] = ldf['conditions'].fillna('Clear')
ldf['conditions_code'] = ldf['conditions'].astype('category').cat.codes

# Drop the old text column
ldf = ldf.drop(columns=['conditions'])


In [ ]:
print(ldf.dtypes)

datetime            datetime64[ns]
tempmax                    float64
tempmin                    float64
temp                       float64
feelslikemax               float64
feelslikemin               float64
feelslike                  float64
dew                        float64
humidity                   float64
precip                     float64
precipprob                 float64
precipcover                float64
snow                       float64
snowdepth                  float64
windgust                   float64
windspeed                  float64
winddir                    float64
sealevelpressure           float64
cloudcover                 float64
visibility                 float64
solarradiation             float64
solarenergy                float64
uvindex                    float64
severerisk                 float64
sunrise             datetime64[ns]
sunset              datetime64[ns]
moonphase                  float64
is_none                       bool
is_rain             

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:

# 1. Classification Target: Will it rain tomorrow? (Shift binary status back 1 step)
ldf['will_rain_tomorrow'] = ldf['is_rain'].shift(-1)

# 2. Time-Series Target: What will the temperature be tomorrow? (Shift continuous temp back 1 step)
ldf['next_day_temp'] = ldf['temp'].shift(-1)

# 3. Clean Boundary Rows: Drop the final row since it has no "tomorrow" to calculate against
ldf = ldf.dropna(subset=['will_rain_tomorrow', 'next_day_temp'])
f = ldf.dropna()

In [ ]:
from sklearn.model_selection import train_test_split

# --- SPLIT A: Classification Matrix ---
# Drop both target columns to protect the input features (X)
X_class = ldf.drop(columns=['will_rain_tomorrow', 'next_day_temp'])
y_class = ldf['will_rain_tomorrow'].astype(int)

# Standard random split (fine for evaluating independent classification events)
X_train_c, X_test_c, y_train_c, y_test_c = train_test_split(
    X_class, y_class, test_size=0.2, random_state=42
)

# --- SPLIT B: Time-Series Matrix ---
X_ts = ldf.drop(columns=['will_rain_tomorrow', 'next_day_temp'])
y_ts = ldf['next_day_temp']

# Chronological split index (Strictly preserves time sequence; no shuffling allowed)
split_idx = int(len(ldf) * 0.8)

X_train_ts = X_ts.iloc[:split_idx]
y_train_ts = y_ts.iloc[:split_idx]
X_test_ts = X_ts.iloc[split_idx:]
y_test_ts = y_ts.iloc[split_idx:]


In [ ]:
# 1. Convert Sunrise and Sunset into pure numeric hours (e.g., 6.25 instead of 06:15 AM)
# This turns timestamps into numbers your model can actually learn from
if 'sunrise' in ldf.columns and ldf['sunrise'].dtype == 'datetime64[ns]':
    ldf['sunrise_hour'] = ldf['sunrise'].dt.hour + (ldf['sunrise'].dt.minute / 60.0)
    ldf = ldf.drop(columns=['sunrise'])

if 'sunset' in ldf.columns and ldf['sunset'].dtype == 'datetime64[ns]':
    ldf['sunset_hour'] = ldf['sunset'].dt.hour + (ldf['sunset'].dt.minute / 60.0)
    ldf = ldf.drop(columns=['sunset'])

# 2. Re-create your feature matrix (X) by dropping ALL remaining datetime columns
# Note: Since 'datetime' is your dataframe index, it won't be harmed here
X_class = ldf.drop(columns=['will_rain_tomorrow', 'next_day_temp'])
y_class = ldf['will_rain_tomorrow'].astype(int)

# 3. Explicitly filter out any stray datetime columns just to be 100% safe
X_class = X_class.select_dtypes(exclude=['datetime64', 'datetime64[ns]', 'object'])

# 4. Re-split your classification data with the clean numeric features
X_train_c, X_test_c, y_train_c, y_test_c = train_test_split(
    X_class, y_class, test_size=0.2, random_state=42
)

# 5. Re-split your time-series data with the clean numeric features
X_ts = ldf.drop(columns=['will_rain_tomorrow', 'next_day_temp']).select_dtypes(exclude=['datetime64', 'datetime64[ns]', 'object'])
y_ts = ldf['next_day_temp']

X_train_ts = X_ts.iloc[:split_idx]
y_train_ts = y_ts.iloc[:split_idx]
X_test_ts = X_ts.iloc[split_idx:]
y_test_ts = y_ts.iloc[split_idx:]

print("Stray datetime columns removed. Features are now purely numeric!")


Stray datetime columns removed. Features are now purely numeric!


In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

# 1. Random Forest Classifier
rf_class_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_class_model.fit(X_train_c, y_train_c)
rf_class_preds = rf_class_model.predict(X_test_c)

print("=== RANDOM FOREST CLASSIFICATION REPORT ===")
print(classification_report(y_test_c, rf_class_preds))

# 2. Linear Classifier (Logistic Regression)
# Note: Weather variables have different ranges; we increase max_iter so the linear optimizer converges safely
lr_class_model = LogisticRegression(max_iter=5000, random_state=42)
lr_class_model.fit(X_train_c, y_train_c)
lr_class_preds = lr_class_model.predict(X_test_c)

print("\n=== LINEAR CLASSIFICATION (LOGISTIC REGRESSION) REPORT ===")
print(classification_report(y_test_c, lr_class_preds))


=== RANDOM FOREST CLASSIFICATION REPORT ===
              precision    recall  f1-score   support

           0       0.84      0.83      0.83       851
           1       0.72      0.72      0.72       501

    accuracy                           0.79      1352
   macro avg       0.78      0.78      0.78      1352
weighted avg       0.79      0.79      0.79      1352


=== LINEAR CLASSIFICATION (LOGISTIC REGRESSION) REPORT ===
              precision    recall  f1-score   support

           0       0.82      0.84      0.83       851
           1       0.71      0.68      0.70       501

    accuracy                           0.78      1352
   macro avg       0.77      0.76      0.76      1352
weighted avg       0.78      0.78      0.78      1352



/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

# Train Random Forest Regressor on sequential historical data
ts_reg_model = RandomForestRegressor(n_estimators=100, random_state=42)
ts_reg_model.fit(X_train_ts, y_train_ts)

# Generate chronological predictions
ts_reg_preds = ts_reg_model.predict(X_test_ts)

# Calculate performance metrics
rmse = np.sqrt(mean_squared_error(y_test_ts, ts_reg_preds))
r2 = r2_score(y_test_ts, ts_reg_preds)

print("=== TIME-SERIES FORECASTING PERFORMANCE ===")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f} degrees")
print(f"R-squared Variance Score (R²): {r2:.2f}")


=== TIME-SERIES FORECASTING PERFORMANCE ===
Root Mean Squared Error (RMSE): 0.86 degrees
R-squared Variance Score (R²): 0.63


In [ ]:
from sklearn.metrics import classification_report, accuracy_score, mean_squared_error, r2_score
import numpy as np

# =====================================================================
# BLOCK 1: CLASSIFICATION RESULTS (Predicting: Will it rain tomorrow?)
# =====================================================================

# 1. Random Forest Classifier Evaluation
rf_class_preds = rf_class_model.predict(X_test_c)
print("=" * 60)
print("             RANDOM FOREST CLASSIFICATION RESULTS             ")
print("=" * 60)
print(classification_report(y_test_c, rf_class_preds, target_names=["No Rain", "Rain"]))
print(f"Overall Accuracy: {accuracy_score(y_test_c, rf_class_preds):.2%}\n")

# 2. Linear Classifier (Logistic Regression) Evaluation
lr_class_preds = lr_class_model.predict(X_test_c)
print("=" * 60)
print("             LINEAR CLASSIFICATION RESULTS                    ")
print("=" * 60)
print(classification_report(y_test_c, lr_class_preds, target_names=["No Rain", "Rain"]))
print(f"Overall Accuracy: {accuracy_score(y_test_c, lr_class_preds):.2%}\n")


# =====================================================================
# BLOCK 2: TIME-SERIES FORECASTING RESULTS (Predicting: Tomorrow's Temp)
# =====================================================================

# 3. Random Forest Regressor Evaluation
ts_reg_preds = ts_reg_model.predict(X_test_ts)
rmse = np.sqrt(mean_squared_error(y_test_ts, ts_reg_preds))
r2 = r2_score(y_test_ts, ts_reg_preds)
mae = np.mean(np.abs(y_test_ts - ts_reg_preds))

print("=" * 60)
print("             TIME-SERIES FORECASTING RESULTS                  ")
print("=" * 60)
print(f"Root Mean Squared Error (RMSE) : {rmse:.2f} degrees")
print(f"Mean Absolute Error (MAE)       : {mae:.2f} degrees")
print(f"R-squared Variance Score (R²)  : {r2:.2f}")
print(f"Model Interpretation            : This model can explain {r2:.1%} of tomorrow's temperature variance.")
print("=" * 60)


             RANDOM FOREST CLASSIFICATION RESULTS             
              precision    recall  f1-score   support

     No Rain       0.84      0.83      0.83       851
        Rain       0.72      0.72      0.72       501

    accuracy                           0.79      1352
   macro avg       0.78      0.78      0.78      1352
weighted avg       0.79      0.79      0.79      1352

Overall Accuracy: 79.07%

             LINEAR CLASSIFICATION RESULTS                    
              precision    recall  f1-score   support

     No Rain       0.82      0.84      0.83       851
        Rain       0.71      0.68      0.70       501

    accuracy                           0.78      1352
   macro avg       0.77      0.76      0.76      1352
weighted avg       0.78      0.78      0.78      1352

Overall Accuracy: 78.03%

             TIME-SERIES FORECASTING RESULTS                  
Root Mean Squared Error (RMSE) : 0.86 degrees
Mean Absolute Error (MAE)       : 0.67 degrees
R-squared Va

1.	Load and explore — how often does it rain in Lagos? Plot rainfall frequency by month
2.	Engineer the target column: create RainTomorrow (1 if next day has rainfall > 0, else 0)
3.	Clean and encode — handle missing weather readings, scale numeric features
4.	Train 3 classifiers — Logistic Regression, Decision Tree, Random Forest
5.	Answer in a markdown cell: “Which weather conditions best predict rain the next day in Lagos, and what months are highest risk?”
6.	Bonus: Plot a monthly rainfall heatmap showing rainy season patterns
Targets & Submission
Field	Details
Success target	F1-score above 0.72 on the rain class
